### PR TITLE
[RISCV][GISEL] legalize, regbankselect, and instruction-select G_PTRMASK

### PR DESCRIPTION
**Pull request by michaelmaitland**
This is done in instruction-select instead of in legalization for the sake of alias analysis.

### CODE REVIEW

**Comment by topperc** [comment]
Is this based off another target? I see AArch64 handles this in instruction selection.

**Comment by michaelmaitland** [comment]
> Is this based off another target? I see AArch64 handles this in instruction selection.

I didn’t see why this opcode couldn’t be handled generally for all targets. That sounded like it fit well where I put it. No reason why it needs to be in a target specific instruction selector. Maybe other targets can eventually reuse the generic logic added here.

**Comment by topperc** [comment]
> > Is this based off another target? I see AArch64 handles this in instruction selection.
> 
> I didn’t see why this opcode couldn’t be handled generally for all targets. That sounded like it fit well where I put it. No reason why it needs to be in a target specific instruction selector. Maybe other targets can eventually reuse the generic logic added here.

If all targets should lower it in the legalizer then why does it exist as an opcode at all?

**Comment by tschuett** [comment]
`G_PTRMASK` is for alias analysis. It should give better results if it exists longer in the pipeline and is only selected at the end. If every target goes through the p->int and int->p early, you will get worse results. Furthermore, GPUs with funny address spaces may do it completely different.

**Comment by michaelmaitland** [comment]
> `G_PTRMASK` is for alias analysis. It should give better results if it exists longer in the pipeline and is only selected at the end. If every target goes through the p->int and int->p early, you will get worse results. Furthermore, GPUs with funny address spaces may do it completely different.

I will move this to instruction selection. Are any targets doing alias analysis between legalization and instruction-selection?

**Comment by tschuett** [comment]
The LoadStoreOpt pass uses AliasAnalysis.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 1]
```diff
@@ -310,6 +310,5 @@ bool RISCVLegalizerInfo::legalizeCustom(LegalizerHelper &Helper,
            LegalizerHelper::Legalized;
   }
   }
-
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:313]: Don't delete this blank line

File: llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/ptrmask-rv32.mir [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/instruction-select/ptrmask-rv32.mir:1]: Should we test the ANDI case which is probably the more common use case for G_PTRMASK?

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**